### PR TITLE
irmin-pack: introduce chunked suffix abstraction

### PR DESCRIPTION
**Pull request by metanivek**
This PR is part of #2096 and introduces a chunked suffix abstraction that works with our current single suffix world. The goal with this change is to introduce the abstraction as a step towards multiple chunks and modifying our GC to work with chunks.

### CODE REVIEW

File: src/irmin-pack/unix/chunked_suffix.ml [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    let open_rw ~root ~chunk_idx ~auto_flush_threshold ~auto_flush_procedure
+        ~end_poff ~dead_header_size =
+      let path = path ~root ~chunk_idx in
+      Ao.open_rw ~path ~end_poff ~dead_header_size ~auto_flush_threshold
+        ~auto_flush_procedure
+
+    let open_ro ~root ~chunk_idx ~end_poff ~dead_header_size ~is_appendable =
+      let open Result_syntax in
+      let path = path ~root ~chunk_idx in
+      let dead_header_size = if is_appendable then dead_header_size else 0 in
```
Comment (icristescu) [inline src/irmin-pack/unix/chunked_suffix.ml:115]: I'm not sure why we need this check. Isn't it the case that `dead_header_size` is 0 for all files created by V3 and above?

**Review by icristescu — APPROVED** [review]
great, LGTM!

File: src/irmin-pack/unix/chunked_suffix.ml (hunk 1 above)
Comment (metanivek) [reply src/irmin-pack/unix/chunked_suffix.ml:115]: I ignored `dead_header_size` at first too but the tests failed (ty tests!). The reason it is needed for the "appendable" (ie last) chunk is that this will be the chunk for migrated v1/v2 stores that have never been (or cannot be -- ie are non-minimal) GCed. The same thing is happening for rw chunks but it is less obvious because there is no check since they are (by definition) "appendable".

**Comment by metanivek** [comment]
Ah, CI is failing because of `Array.find_opt` not being available on older OCaml versions -- will fix this.

File: src/irmin-pack/unix/chunked_suffix.ml (hunk 1 above)
Comment (metanivek) [reply src/irmin-pack/unix/chunked_suffix.ml:115]: Aha! You're right. Previously, I was always using `dead_header_size=0`, which broke tests, but I confused myself when I applied the fix. Just using the passed in `dead_header_size` does work. Thanks for the investigation!

File: src/irmin-pack/unix/chunked_suffix.ml [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    let iter_skip_appendable f t =
+      Array.sub t 0 (Array.length t - 1) |> Array.iter f
+
+    let appendable t = Array.get t (Array.length t - 1)
+
+    let find ~off t =
+      let find c =
+        let end_poff = Ao.end_poff c.ao in
+        let is_after_start = c.start_off <= off in
+        let is_before_end = chunk_off_to_poff c off < end_poff in
```
Comment (art-w) [inline src/irmin-pack/unix/chunked_suffix.ml:69]: Do we need to use the Int63 comparisons to avoid a surprise on non-64-bits platforms?

File: src/irmin-pack/unix/chunked_suffix.ml [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    { inventory }
+
+  let open_ro ~root ~end_poff ~dead_header_size ~start_idx ~chunk_num =
+    let open Result_syntax in
+    let open_chunk = Ao_chunk.open_ro ~root ~end_poff ~dead_header_size in
+    let+ inventory = Inventory.open_ ~start_idx ~chunk_num ~open_chunk in
+    { inventory }
+
+  let appendable_ao t = (Inventory.appendable t.inventory).ao
+  let end_poff t = appendable_ao t |> Ao.end_poff
```
Comment (icristescu) [inline src/irmin-pack/unix/chunked_suffix.ml:159]: I think this is wrong -  the `end_poff` is used upstreamed as the end poff of the whole suffix abstraction, but here it returns only the end poff of the last chunk. I should be the sum over all chunks.

File: src/irmin-pack/unix/chunked_suffix.ml [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    (* Close immutable chunks, ignoring errors. *)
+    let _ =
+      t.inventory
+      |> Inventory.iter_skip_appendable @@ fun chunk ->
+         let _ = Ao.close chunk.ao in
+         ()
+    in
+    (* Close appendable chunk and keep error since this
+       is the one that can have a pending flush. *)
+    appendable_ao t |> Ao.close
```
Comment (art-w) [inline src/irmin-pack/unix/chunked_suffix.ml:178]: Nitpick: Perhaps it would be easier to implement the `close` in the `Inventory` module rather than expose an `iter_skip_appendable`?

File: src/irmin-pack/unix/chunked_suffix.ml [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+    let v = Array.init
+
+    let iter_skip_appendable f t =
+      Array.sub t 0 (Array.length t - 1) |> Array.iter f
+
+    let appendable t = Array.get t (Array.length t - 1)
+
+    let find ~off t =
+      let find c =
+        let end_poff = Ao.end_poff c.ao in
```
Comment (icristescu) [inline src/irmin-pack/unix/chunked_suffix.ml:67]: again, I don't think this `Ao.end_poff` is want you should use here

If I understand correctly, this abstraction introduces a new layer of offsets. We now have:
- global offsets (dispatched between prefix and suffix);  
- suffix offsets (physical offsets in the old suffix)  -- the one you manipulate in most of this file, right?
- physical chunk offsets (the one returned by `Ao.end_poff` for instance).

File: src/irmin-pack/unix/chunked_suffix.ml [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+    let open_chunk = Ao_chunk.open_ro ~root ~end_poff ~dead_header_size in
+    let+ inventory = Inventory.open_ ~start_idx ~chunk_num ~open_chunk in
+    { inventory }
+
+  let appendable_ao t = (Inventory.appendable t.inventory).ao
+  let end_poff t = appendable_ao t |> Ao.end_poff
+
+  let read_exn t ~off ~len buf =
+    let chunk = Inventory.find ~off t.inventory in
+    let poff = chunk_off_to_poff chunk off in
```
[the remark this answers is not in the document]
Comment (art-w) [reply src/irmin-pack/unix/chunked_suffix.ml:163]: I actually like that a chunked file is an alternative abstraction for implementing an append-only file, it looks very clean not to leak the chunking trick elsewhere.
However, I would prefer if the abstraction was not in charge of choosing the chunk filenames (for example by receiving the list of chunk filenames as an argument to `open_*` from the file manager) But we could clean this in a parallel PR in the interest of having @metanivek move forward with the MVP!

File: src/irmin-pack/unix/chunked_suffix.ml (hunk 5 above)
Comment (art-w) [reply src/irmin-pack/unix/chunked_suffix.ml:67]: I think this one is fine because it's subtracting the chunk `start_off` in `chunk_off_to_poff` before comparing (?)

Comment (icristescu) [reply src/irmin-pack/unix/chunked_suffix.ml:67]: Yes, just saw that, thanks!

File: src/irmin-pack/unix/chunked_suffix.ml (hunk 6 above)
[the remark this answers is not in the document]
Comment (Ngoguey42) [reply src/irmin-pack/unix/chunked_suffix.ml:163]: > I actually like that a chunked file is an alternative abstraction for implementing an append-only file, it looks very clean not to leak the chunking trick elsewhere.

You've pinpointed the great advantage of this approach: a patch that feels natural and that moves the minimal amount of dirt. This seems like the the best approach right now, but...

When you say "elsewhere" you are in fact referring to the dispatcher/file_manager pair that are abstractions under our control that have been written very recently for that exact purpose. 

The situation on master is: The file_manager holds the file descriptor and the dispatcher chooses a file descriptor and an offset. Cheating with that [KISS](https://en.wikipedia.org/wiki/KISS_principle) fact will increase the cognitive complexity of irmin-pack. If we preserve that situation I am confident that we will have no surprise when implementing the lower layer.

To illustrate that, picture a new comer to the codebase that is on the path to learning how irmin-pack updates and syncs with the file-system:
- They will find a dispatcher abstraction that doesn't take care of all the dispatching.
- On top of the global offsets and the physical offset they will find a third type of offset which corresponds to the conversion from global to legacy-suffix, just before the conversion from legacy-suffix to physical.

[the remark this answers is not in the document]
Comment (metanivek) [reply src/irmin-pack/unix/chunked_suffix.ml:163]: @Ngoguey42 thanks for the feedback and historical perspective. I did consider putting this code across the File Manager and Dispatcher but (as is evidenced by this PR :smile:) decided against that path. Before responding to some of the above points, here is more context on my thought process.

I think it is important for us to capture the conceptual pieces of software as self-contained modules with good interfaces. Spreading the concept/logic/usage for a chunked suffix across two (or likely more when you involve `Pack_store`'s usage of `Suffix`) modules does not feel optimal to me. This is probably the most important and motivating reason for me creating a `Chunked_suffix` module. Why is this important to me? I believe that identifying the important concepts and modelling them with good interfaces is one of the most important tasks when trying to build maintainable software that can change over time. I think chunked suffix is important enough that it has its own interface and module.

As I look at the two files mentioned, File Manager and Dispatcher, here is how I see the distinction and how I think a little bit about how this code might evolve in the future:

- The File Manager is the container/orchestrating entity for the conceptual files (Control, Dict, Prefix, Suffix, Index, etc.) that compose an `irmin-pack` store. It is responsible for opening, closing, flushing, swapping, etc.. It also does some reading and writing of these files but: 
	- reading from "the pack file" is coordinated by the Dispatcher 
	- writing also happens directly to the Suffix (eg in `Pack_store`).
- The Dispatcher is the router for reading from the pack file (verbosely Dispatcher is `Pack_file_router` in my mind). Today, the pack file conceptually is a Prefix and a Suffix (independently of how the suffix is implemented). This PR does not change the routing logic between these two concepts. This is one of the strengths (from my perspective) of the approach -- the PR merely replaces how the concept of Suffix is implemented with minimal changes to how you create or interact with it. In the future, I could see us adding a `Vault`/`Lower`/`Whatever` concept for the archive store that the Dispatcher is also responsible for routing to when reading data.

---

I think the above addresses some of the comments but I'd like to reiterate/respond directly, too.

> If we preserve that situation I am confident that we will have no surprise when implementing the lower layer.
For a similar motivating reason as `Chunked_suffix`, I think the lower layer is an important enough concept that it will have its own interface/module and then plug into the Dispatcher's routing logic. I don't foresee any difficulties/surprises with this but I'd be interested if you do.

> a new comer to the codebase that is on the path to learning how irmin-pack updates and syncs with the file-system
This is a difficult code base to comprehend and in part I think (I haven't done a detailed analysis this is more of a hunch :salt:) because concepts/details tend to leak across module boundaries. I think File Manager and Dispatcher are a good step towards adding important concepts to the code base. Similar to the work to break apart the Gc module into good stand-alone concepts (and not a junk drawer utils :wink:), I think adding important concepts (like chunked suffix) as stand-alone modules will benefit the code over time and help paths of learning.

> They will find a dispatcher abstraction that doesn't take care of all the dispatching.
The Dispatcher routes between prefix and suffix. It is true the the chunked suffix has its own internal "dispatching" but I don't find this to be a problem since it is in the scope of the chunked suffix itself which makes it easy to reason about and explain (from my perspective).

> On top of the global offsets and the physical offset they will find a third type of offset which corresponds to the conversion from global to legacy-suffix, just before the conversion from legacy-suffix to physical.
Chunks require a new offset mapping, regardless of where it is implemented. The way this PR maps offsets (and how I would think about it even if the code were located say in the Dispatcher) is global offset -> suffix offset -> chunk physical offset. What you are calling `legacy-suffix` is just the conceptual suffix (and perhaps upper suffix in the future). You could maybe avoid the suffix offset if this code lived in the Dispatcher, but I think there is strength and clarity in its presence:
- it fits and reinforces the model of GC splitting the pack file into a prefix and suffix
- when considering `Chunked_suffix` code you only need to think about suffix offset to chunk off/poff mapping 

---

Hopefully the above makes sense and at least gives reason for why I chose the path I did and how I'm considering future changes.

**tl;dr** I think adding important concepts like Chunked Suffix as stand-alone modules with well-defined interfaces is a good architectural decision and will benefit the code as we continue to evolve and improve it. I also think it fits into the current architecture of File Manager and Dispatcher at the conceptual level.

File: src/irmin-pack/unix/chunked_suffix.ml (hunk 3 above)
Comment (metanivek) [reply src/irmin-pack/unix/chunked_suffix.ml:159]: Thanks for bringing this up. I think this is worth some discussion.

The way I understand how the suffix's `poff` is currently used in the control file is as a control mechanism to know/verify what data we have written to disk (as is seen in the append only file's consistency check). In this sense, the current code is correct. It tracks the physical offset of the appendable chunk for consistency checks. No other chunk can be changed so its `end_poff` is necessarily equivalent to its `Io.size_of_path` (the awkward code in `Ao_chunk.open_ro` that does exactly this).

Summing the values would give the chunked suffix's length and final offset but doesn't really correlate with a physical offset

While working on this I intentionally wanted to keep the existing code working with minimal extra changes but I do think we need to change this. Here is my proposal for a future PR:

- rename control file field to `suffix_consistency_poff` (or something else -- open to ideas!)
- rename `end_poff` in chunked file to `consistency_poff`
- move consistency checking up from append only to chunked suffix and remove `end_poff` from append only's api

File: src/irmin-pack/unix/chunked_suffix.ml (hunk 4 above)
Comment (metanivek) [reply src/irmin-pack/unix/chunked_suffix.ml:178]: I iterated a bit on the boundary between `Inventory` and the encompassing module. The current distinction that I settled upon is that `Inventory` is for storing/looking up chunks (with light concepts like appendable for easier access) but doesn't call functions on `Ao` (with the exception of `end_poff` since that is a property used for finding chunks). I wanted any interactions with the `Ao.t` to be in the domain of the encompassing module.

I am definitely open to exploring this boundary though! Maybe see how subsequent changes go and revise later?

File: src/irmin-pack/unix/chunked_suffix.ml (hunk 5 above)
Comment (metanivek) [reply src/irmin-pack/unix/chunked_suffix.ml:67]: @icristescu your breakdown of the mapping of offsets is correct. A `Chunked_suffix` maps suffix offset to chunk physical offset (which each chunk storing its suffix offset to facilitate lookup).

Based on this discussion, I think the names could be improved:

- `chunk.start_off` could be `chunk.suffix_off`
- `chunk_off_to_poff` could be `suffix_off_to_chunk_poff` (verbose but clear (?))

WDYT?

File: src/irmin-pack/unix/chunked_suffix.ml (hunk 6 above)
[the remark this answers is not in the document]
Comment (metanivek) [reply src/irmin-pack/unix/chunked_suffix.ml:163]: @art-w 

> However, I would prefer if the abstraction was not in charge of choosing the chunk filenames

I like this idea. I wasn't quite sure about passing the root. To add to your idea of passing a list of path names: instead of `root : string` the open/create functions could have a `chunk_path : (chunk_idx:int -> string)` parameter. This leaves control with chunked suffix for creating the paths it needs but removes from its control where those paths are. That type would also play nicely with `Layout.V4.suffix_chunk ~root`.

[the remark this answers is not in the document]
Comment (art-w) [reply src/irmin-pack/unix/chunked_suffix.ml:163]: When I first read the codebase, I missed a similar abstraction for "sparse file" to clarify that the mapping file is only used with a compacted object file to implement a "read-only file" interface (and that the mapping doesn't overlap with the suffix). Currently this is conflated in the dispatcher and I think it doesn't scale to also have the chunks logic in there... But the dispatcher could only be in charge of dispatching left to the sparse prefix file, or right to the chunked suffix based on the split offset.

File: src/irmin-pack/unix/chunked_suffix.ml (hunk 2 above)
Comment (metanivek) [reply src/irmin-pack/unix/chunked_suffix.ml:69]: Oh yikes. Yes! Good catch. TIL that this is an area where I need to not trust the OCaml compiler to yell at me. :sweat_smile:

File: src/irmin-pack/unix/chunked_suffix.ml (hunk 6 above)
[the remark this answers is not in the document]
Comment (art-w) [reply src/irmin-pack/unix/chunked_suffix.ml:163]: Hm well, a `string list` or a `chunk_num:int * chunk_path:(int -> string)` are isomorphic for your purpose :P but the list seems easier to read in the signature and with no "out of bounds" risk! (This is only shifting a `List.init` to the file manager)

[the remark this answers is not in the document]
Comment (metanivek) [reply src/irmin-pack/unix/chunked_suffix.ml:163]: It is a bit like comparing direct array/list construction with `(Array|List).init`. As a hunch I prefer the latter pattern here to leave control with the chunked suffix, but I need to think more about why I have this preference. I think I could likely be happy with either change. I also have a hunch that this particular point might be clear how to resolve once we have split and gc in place since we will have a complete picture of how the chunked suffix works. Good news: we have two isomorphic iterations to entertain! :smile:

File: src/irmin-pack/unix/chunked_suffix.ml (hunk 5 above)
Comment (metanivek) [reply src/irmin-pack/unix/chunked_suffix.ml:68]: Made these name changes in [`c55a987` (#2115)](https://github.com/mirage/irmin/pull/2115/commits/c55a98774e1741a41021f6619e3581c2abf8ac66) and I think it improves the clarity!

File: src/irmin-pack/unix/chunked_suffix.ml (hunk 6 above)
[the remark this answers is not in the document]
Comment (icristescu) [reply src/irmin-pack/unix/chunked_suffix.ml]: (I'm adding my two cents here, but with no intention of blocking this PR, feel free to ignore :) )

I agree with @Ngoguey42 on this, I would much prefer to have the dispatching logic between files be entirely in the dispatcher. There are several bits of logic that I find hard to track, and I think we should try to have them in a single place: 
- physical/virtual offsets conversions;
- reads/writes that lead to errors (due to file boundaries)
- converting read/write to IO.read/write (more abstractions we have, more indirection too).

> Spreading the concept/logic/usage for a chunked suffix across two (or likely more when you involve Pack_store's usage of Suffix) 

It should be that all code above the dispatcher use the dispatcher's read and write, the fact that the suffix leaks in the pack store, I think is something to fix.

File: src/irmin-pack/unix/chunked_suffix.ml (hunk 3 above)
Comment (icristescu) [reply src/irmin-pack/unix/chunked_suffix.ml:159]: The dispatcher needs to know the end offset for the suffix file in order to verify that a read is within bounds.  This does not correspond to a physical offset, indeed. It's "a virtual offset corresponding to the physical end offset, if the suffix was a single file". I'll call it `suffix_end_off`.

We can either (I am reiterating what you say):
- keep `end_poff` as is, keep the control file as is (tracking the `end_poff` of the last chunk), but then we need a function to compute the  `suffix_end_off` - solution works for me, I do find the concept of `suffix_end_off` a bit weird. 
- replace the `end_poff` in the chunked file and the control file with the `suffix_end_off` - I'm not sure if this is what you are proposing for a future PR?

Comment (metanivek) [reply src/irmin-pack/unix/chunked_suffix.ml:159]: Ah, I just looked in the dispatcher and see its usage of `Suffix.end_poff` -- this does need to be updated to be the suffix end offset. Before chunks, this makes sense as an `end_poff`, but there is no "end poff" for a chunked suffix. I don't necessarily love adding another virutalized set of offsets, but the suffix's offset range will necessarily be virtualized for chunks since it doesn't map directly to a single range of physical offsets any longer. And it is neatly contained within `Chunked_suffix` which helps reduce the scope of offsets to consider.

Thanks for pointing out dispatcher's use of the end offset for the suffix. I think the following proposal addresses everything:

- Add `end_off` to `Chunked_suffix` to represent the last offset of the suffix. This can be a sum of the chunks poffs but I don't think tracking it in the control file is needed.
- Rename `Chunked_suffix.end_poff` to `consistency_poff` (or something else -- but I think "end poff" doesn't make sense in the context of a chunked suffix) and also the control file field to reflect the name change. This is the offset we need to track in the control file for consistency checks since we just need to check the appendable chunk's poff upon open. As a part of this, also move the consistency check from append only to chunked suffix (like mentioned previously).

I will make the first change in this PR since you rightly pointed out a correctness issue. The second one can be a future PR since it is mostly a cosmetic name change, but it will also get rid of some of the awkward `end_poff` code (like in `Ao_chunk.open_ro`)  which will be nice.

Comment (metanivek) [reply src/irmin-pack/unix/chunked_suffix.ml:159]: Aha, I missed that. I'll just review in your PR since you have already made the change. :+1:

**Comment by icristescu** [comment]
@metanivek can you add a changelog entry to mention that the file disk format changed to chunk files?

File: src/irmin-pack/unix/chunked_suffix.ml (hunk 6 above)
[the remark this answers is not in the document]
Comment (Ngoguey42) [reply src/irmin-pack/unix/chunked_suffix.ml]: We discussed this offline with @metanivek. I would TL;DR our discussion this way: I believe that what lands on the main branch should be a somewhat definitive version of the code that everyone agrees on while your approach is more iterative.

Let's do it your way.

File: src/irmin-pack/unix/chunked_suffix.ml [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+    (* Close appendable chunk and keep error since this
+       is the one that can have a pending flush. *)
+    appendable_ao t |> Ao.close
+
+  let empty_buffer t = appendable_ao t |> Ao.empty_buffer
+  let flush t = appendable_ao t |> Ao.flush
+  let fsync t = appendable_ao t |> Ao.fsync
+
+  let refresh_end_poff t new_end_poff =
+    Ao.refresh_end_poff (appendable_ao t) new_end_poff
```
[the remark this answers is not in the document]
Comment (icristescu) [reply src/irmin-pack/unix/chunked_suffix.ml:185]: My understanding is that this the end poff for the last chunk and not the length of the full suffix. The control file also tracks the end poff for the last chunk. 
We only need the length of the full suffix in the dispatcher, for which an additional function is needed. I suggested to add it in the my PR https://github.com/mirage/irmin/pull/2115#discussion_r997138686 but we can add it here if you prefer (I don't think it matters much, since there is no split and no tests here).

File: src/irmin-pack/unix/chunked_suffix.ml [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+  module Ao_chunk = struct
+    let path = Layout.V4.suffix_chunk
+
+    let create_rw ~root ~chunk_idx ~overwrite ~auto_flush_threshold
+        ~auto_flush_procedure =
+      let path = path ~root ~chunk_idx in
+      Ao.create_rw ~path ~overwrite ~auto_flush_threshold ~auto_flush_procedure
+
+    let open_rw ~root ~chunk_idx ~auto_flush_threshold ~auto_flush_procedure
+        ~end_poff ~dead_header_size =
```
[the remark this answers is not in the document]
Comment (metanivek) [reply src/irmin-pack/unix/chunked_suffix.ml:107]: Aha, you're right -- good catch. This code works with the current PR because there is no split. I'm happy to rework this in this PR or let @icristescu take it in the split PR.

File: src/irmin-pack/unix/chunked_suffix.ml (hunk 4 above)
Comment (metanivek) [reply src/irmin-pack/unix/chunked_suffix.ml:178]: @Ngoguey42 okay, that discrepancy convinces me to move it. I cannot unsee that now.

File: src/irmin-pack/unix/chunked_suffix.ml (hunk 7 above)
[the remark this answers is not in the document]
Comment (metanivek) [reply src/irmin-pack/unix/chunked_suffix.ml:185]: Yeah, we should keep track of these somehow, either TODOs or as a comment with checkboxes in the main issue. I'll do something to help us remember.

Adding to what @icristescu says: for `end_poff`, as mentioned in https://github.com/mirage/irmin/pull/2115#discussion_r995922877, I do think we ought to rename it in chunked suffix and in control file (`consistency_poff`, `persisted_poff`, ???) to try to make its purpose clearer. I think a pair of functions, `Chunked_suffix.consistency_poff` and `Chunked_suffix.refresh_consistency_poff`, would make it clearer, even if they are just delegating to the last chunk's `end_poff`.

As for the length of the suffix, I am okay to have it in the split PR (with a suggestion for a rename :smile: https://github.com/mirage/irmin/pull/2118#discussion_r998604496).

File: src/irmin-pack/unix/chunked_suffix.ml [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+        match open_result with
+        | Error err -> raise (OpenInventoryError err)
+        | Ok ao ->
+            let chunk_len = Ao.end_poff ao in
+            (off_acc := Int63.Syntax.(suffix_off + chunk_len));
+            { idx = chunk_idx; suffix_off; ao }
+      in
+      try Ok (v chunk_num create_chunk)
+      with OpenInventoryError err ->
+        Error (err : open_error :> [> open_error ])
```
[the remark this answers is not in the document]
Comment (metanivek) [reply src/irmin-pack/unix/chunked_suffix.ml:92]: Yeah, I thought about that. I don't generally like exceptions as control flow, but I do like the directness of the approach here (and its nice to not _have_ to write the result handling loop code even though it is simple :smile:). That said, I don't feel strongly either way. Do you have a 51% preference to change this?

[the remark this answers is not in the document]
Comment (Ngoguey42) [reply src/irmin-pack/unix/chunked_suffix.ml:92]: This is a very classic monad workflow. I really feel that the exception is superfluous here

File: src/irmin-pack/unix/chunked_suffix.ml (hunk 8 above)
[the remark this answers is not in the document]
Comment (metanivek) [reply src/irmin-pack/unix/chunked_suffix.ml]: I just [fixed this](https://github.com/mirage/irmin/compare/43929d1b048439db9c65ce4ab820757b3e4fdfd0..8f2229f96581fc3dcd742869770c868268bd5825) for this PR (and removed the `Ao_chunk` module in the process :tada:!)

File: src/irmin-pack/unix/chunked_suffix.ml [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+    let+ inventory = Inventory.open_ ~start_idx ~chunk_num ~open_chunk in
+    { inventory }
+
+  let appendable_ao t = (Inventory.appendable t.inventory).ao
+  let end_poff t = appendable_ao t |> Ao.end_poff
+
+  let read_exn t ~off ~len buf =
+    let chunk = Inventory.find ~off t.inventory in
+    let poff = suffix_off_to_chunk_poff chunk off in
+    Ao.read_exn chunk.ao ~off:poff ~len buf
```
[the remark this answers is not in the document]
Comment (metanivek) [reply src/irmin-pack/unix/chunked_suffix.ml]: Instead of an assert, I [changed](https://github.com/mirage/irmin/compare/43929d1b048439db9c65ce4ab820757b3e4fdfd0..8f2229f96581fc3dcd742869770c868268bd5825) `find` to return a `poff` and moving the conversion code into that function as well. If the reader is unconvinced, they can look in the `find` code, which should give everything they need.

File: src/irmin-pack/unix/chunked_suffix.ml [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+      let path = path ~root ~chunk_idx in
+      (* Passing [end_poff] is required by Ao for its consistency check (and
+         only applies to the appendable chunk).
+         TODO: We could change its API in the future so we can not do
+         [Io.size_of_path], which is admittedly very silly.*)
+      let* end_poff =
+        if is_appendable then Ok end_poff else Io.size_of_path path
+      in
+      Ao.open_ro ~path ~end_poff ~dead_header_size
+  end
```
[the remark this answers is not in the document]
Comment (metanivek) [reply src/irmin-pack/unix/chunked_suffix.ml]: Ended up [removing this](https://github.com/mirage/irmin/compare/43929d1b048439db9c65ce4ab820757b3e4fdfd0..8f2229f96581fc3dcd742869770c868268bd5825) as a part of addressing a different [comment](https://github.com/mirage/irmin/pull/2115#discussion_r999244077). Now the specifics for mapping from chunk to append-only file are contained in a focused module, `Ao_shim`.

File: src/irmin-pack/unix/chunked_suffix.ml (hunk 9 above)
[the remark this answers is not in the document]
Comment (metanivek) [reply src/irmin-pack/unix/chunked_suffix.ml:96]: Adding this to the list of future clean ups/considerations.

File: src/irmin-pack/unix/chunked_suffix.ml (hunk 4 above)
Comment (metanivek) [reply src/irmin-pack/unix/chunked_suffix.ml]: [Made the change](https://github.com/mirage/irmin/compare/43929d1b048439db9c65ce4ab820757b3e4fdfd0..8f2229f96581fc3dcd742869770c868268bd5825)!

**Comment by metanivek** [comment]
This is ready for a final review before merging. I have resolved all of the comments that I have addressed in the PR. I will add the list of things to revisit to #2096 once this merges.

File: src/irmin-pack/unix/chunked_suffix.ml [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+      in
+      match Array.find_opt find t with
+      | None -> raise (Errors.Pack_error `Read_out_of_bounds)
+      | Some c -> (c, suffix_off_to_chunk_poff c)
+
+    let open_ ~start_idx ~chunk_num ~open_chunk =
+      let off_acc = ref Int63.zero in
+      let create_chunk i =
+        let suffix_off = !off_acc in
+        let is_first = i = 0 in
```
[the remark this answers is not in the document]
Comment (metanivek) [reply src/irmin-pack/unix/chunked_suffix.ml]: Changed to a `is_legacy` flag: https://github.com/mirage/irmin/compare/8f2229f96581fc3dcd742869770c868268bd5825..a1204da2048260384b9e9ac86b10b7883166c19b

**Review by Ngoguey42 — APPROVED** [review]
Thanks. Ship it!
